### PR TITLE
[Serializer] Fix MissingConstructorArgumentsException returning missing argument one by one

### PR DESCRIPTION
**Pull request by ktherage**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR is an improvment of #49013 and fix #42712.

She aims to fix the problem reported in #49013 indicating that the method `getMissingConstructorArguments()` of `\Symfony\Component\Serializer\Exception\MissingConstructorArgumentsException` class introduced in #42712 is returning only one missing arguments when there's more than one missing.

### CODE REVIEW

File: .github/expected-missing-return-types.diff [hunk 1]
```diff
@@ -113,7 +113,7 @@ index bf68ec0dcc..509e5f1b0c 100644
 +++ b/src/Symfony/Bridge/Doctrine/DependencyInjection/CompilerPass/RegisterEventListenersAndSubscribersPass.php
 @@ -54,5 +54,5 @@ class RegisterEventListenersAndSubscribersPass implements CompilerPassInterface
      }
-
```
Comment (WebMamba) [inline .github/expected-missing-return-types.diff]: Looks like those changes are unrelated to this PR

**Comment by HypeMC** [comment]
The original goal of my PR was to add the class as an argument to the exception, this reverts that as well. While I like the idea of having all the arguments at once, I don't see why the part with the class being passed needs to be removed.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 2]
```diff
@@ -290,14 +287,9 @@ abstract protected function extractAttributes(object $object, string $format = n
 
     /**
      * Gets the attribute value.
-     *
-     * @return mixed
```
Comment (WebMamba) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: the return type should not be remove here

Comment (ktherage) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: it's a fabbot.io fix and it seems explicit to me to expect a "mixed" when it's not typed IMO. So the `@return` mixed is not needed here IMO.

Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: this is a false positive from fabbot: we add the annotation to let ppl know that we're going to add a native return type in v7.0

**Comment by ktherage** [comment]
@HypeMC

> The original goal of my PR was to add the class as an argument to the exception, this reverts that as well.

Well, sorry, I missed that part I'll check that.

**Comment by ktherage** [comment]
I have a failing test locally and I think that it's fine as it requires now a constructor parameter that is missing but I'm not sure if I might fix it (ie I actually broke a thing).

It's at src/Symfony/Component/Serializer/Tests/SerializerTest.php:764

details : 
```
1) Symfony\Component\Serializer\Tests\SerializerTest::testUnionTypeDeserializableWithoutAllowedExtraAttributes
Symfony\Component\Serializer\Exception\MissingConstructorArgumentsException: Cannot create an instance of "Symfony\Component\Serializer\Tests\DummyCTypeForUnion" from serialized data because its constructor requires the following parameters to be present : "c".

/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php:408
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:244
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:323
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Serializer.php:227
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:537
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:601
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php:370
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:244
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:323
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Serializer.php:227
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Serializer.php:145
/home/ktherage/Documents/forks/symfony/src/Symfony/Component/Serializer/Tests/SerializerTest.php:764
```

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php [hunk 3]
```diff
@@ -12,37 +12,61 @@
...
+     * @var MissingConstructorArgumentException[]
      */
-    private $missingArguments;
+    private $missingConstructorArgumentExceptions = [];
 
-    public function __construct(string $message, int $code = 0, \Throwable $previous = null, array $missingArguments = [])
+    /**
+     * @param MissingConstructorArgumentException[]     $missingArguments
+     */
+    public function __construct(array $missingArguments)
```
Comment (HypeMC) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: Changing the constructor signature is a BC break, why not just add a new optional argument:

```php
public function __construct(string $message, int $code = 0, \Throwable $previous = null, array $missingArguments = [], string $class = null)
```

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php [hunk 4]
```diff
@@ -12,37 +12,66 @@
...
      */
-    private $missingArguments;
+    private $missingConstructorArgumentExceptions = [];
 
-    public function __construct(string $message, int $code = 0, \Throwable $previous = null, array $missingArguments = [])
+    /**
+     * @param ?class-string $class
+     * @param MissingConstructorArgumentException[] $missingArguments
+     */
+    public function __construct(string $message, int $code = 0, \Throwable $previous = null, string $class = null, array $missingArguments = [])
```
Comment (HypeMC) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: This is still a BC break, any new argument needs to be added to the end so that existing code wouldn't break, eg:

```php
new MissingConstructorArgumentsException('', 0, null, []); // this would now break
```

Also, I'd suggest removing the `MissingConstructorArgumentException` as it's no longer needed.

Comment (ktherage) [reply src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: OK, I'll set back the parameter order.

> Also, I'd suggest removing the MissingConstructorArgumentException as it's no longer needed.

I think that having both could be a nice to have in the case that you might need to iterate over exception or maybe adapt serialization for them.

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php [hunk 5]
```diff
@@ -12,37 +12,66 @@
...
+        if (count($classes) > 1) {
+            throw new \InvalidArgumentException(sprintf(
+                'All instances of "%s" must concern the same class. Classes given : "%s".',
+                MissingConstructorArgumentException::class,
+                implode('", "', $classes)
+            ));
         }
-
-        $this->missingArguments = $missingArguments;
+        if ($class !== null && $class !== $classes[0]) {
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: It'll fail if `$classes` is `[]` (if for example the given `$missingArguments` is empty)
You can maybe replace `$classes[0]` by `($classes[0] ?? null)`.

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php [hunk 6]
```diff
@@ -12,37 +12,66 @@
...
+            throw new \InvalidArgumentException(sprintf(
+                'The class given in the constructor "%s" must be the same as the one given in the "%s" instances which is "%s".',
+                $class,
+                MissingConstructorArgumentException::class,
+                $classes[0]
+            ));
+        }
+        $this->class = $class ?: $classes[0];
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: Same here
```suggestion
        $this->class = $class ?: ($classes[0] ?? null);
```

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 7]
```diff
@@ -331,9 +332,9 @@ protected function instantiateObject(array &$data, string $class, array &$contex
                 return $reflectionClass->newInstanceWithoutConstructor();
             }
 
+            $context['missing_constructor_arguments'] = [];
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Why is it set in the context here? Can't it be a variable?

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php (hunk 5 above)
Comment (ktherage) [reply src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: Nice catch.

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php (hunk 6 above)
Comment (ktherage) [reply src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: This shall never be the case here IMO as the above if checks it and throws an exception.
Am I missing something ?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 7 above)
Comment (ktherage) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I made it like this to keep the same "code philosophy" as for `not_normalizable_value_exceptions` on line 401.

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php [hunk 8]
```diff
@@ -12,37 +12,66 @@
...
+    /**
+     * @param MissingConstructorArgumentException[] $missingArguments
+     * @param ?class-string $class
+     */
+    public function __construct(string $message, int $code = 0, \Throwable $previous = null, array $missingArguments = [], string $class = null)
     {
-        if (!$this instanceof MissingConstructorArgumentException) {
-            trigger_deprecation('symfony/serializer', '6.3', 'The "%s" class is deprecated, use "%s" instead.', __CLASS__, MissingConstructorArgumentException::class);
+        $this->missingConstructorArgumentExceptions = $missingArguments;
+        $classes = array_unique(array_map(fn (MissingConstructorArgumentException $missingConstructorArgumentException) => $missingConstructorArgumentException->getClass(), $missingArguments));
```
Comment (HypeMC) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: This is still a breaking change, any existing code that passes an array of strings will fail.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 7 above)
Comment (mtarld) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Yes, but it's because the `not_normalizable_value_exceptions` is used as a reference elsewhere in the code (see `src/Symfony/Component/Serializer/Serializer.php`)

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php (hunk 6 above)
Comment (mtarld) [reply src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: If `$class` is null and `$classes` is `[]`, we won't enter the upper if, and won't throw any exception, right?

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php [hunk 9]
```diff
@@ -12,37 +12,66 @@
...
+        if (count($classes) > 1) {
+            throw new \InvalidArgumentException(sprintf(
+                'All instances of "%s" must concern the same class. Classes given : "%s".',
+                MissingConstructorArgumentException::class,
+                implode('", "', $classes)
+            ));
         }
-
-        $this->missingArguments = $missingArguments;
+        if ($classes[0] ?? null) {
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: Why dropping the `$class` comparison?

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php (hunk 8 above)
Comment (ktherage) [reply src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: I've reworked that and added test to be sure that both options are allowed but strings shall trigger a deprecation now.

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php (hunk 9 above)
Comment (ktherage) [reply src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentsException.php]: Sorry, I've reworked that based on HypeMC's BC Break notice.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 7 above)
Comment (ktherage) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: ok, I supposed it was some kind of "convention". I've fixed that.

File: src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentException.php [hunk 10]
```diff
@@ -11,31 +11,31 @@
 
 namespace Symfony\Component\Serializer\Exception;
 
-class MissingConstructorArgumentException extends MissingConstructorArgumentsException
+class MissingConstructorArgumentException extends RuntimeException
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Exception/MissingConstructorArgumentException.php]: can't we get rid of this class altogether and keep only `MissingConstructorArgumentsException`
also: please mind fabbot ;)

**Comment by nicolas-grekas** [comment]
Thank you @ktherage.
